File: windows/src/GoalMaker.Core/Updates/UpdatePolicy.cs
using GoalMaker.Core.Versioning;

namespace GoalMaker.Core.Updates;

/// <summary>
/// Decides whether an available release is offered to the installed app. Development builds never
/// auto-update and pre-releases are never offered (contracts/vectors/semantic-version.json).
/// </summary>
public static class UpdatePolicy
{
    public static bool ShouldOffer(string installed, string available)
    {
        var current = SemanticVersion.Parse(installed);
        var candidate = SemanticVersion.Parse(available);
        if (current is null || candidate is null || current.IsDevelopmentBuild || candidate.IsPrerelease)
        {
            return false;
        }

        return candidate > current;
    }
}

File: android/app/src/main/kotlin/com/goalmaker/app/domain/update/UpdatePolicy.kt
package com.goalmaker.app.domain.update

import com.goalmaker.app.domain.version.SemanticVersion

/**
 * Decides whether an available release should be offered to the installed app. A development build
 * never auto-updates (CodePrint delivery rule), and pre-releases are never offered.
 * Fixed by the offerUpdate cases in contracts/vectors/semantic-version.json.
 */
object UpdatePolicy {
    fun shouldOffer(installed: String, available: String): Boolean {
        val current = SemanticVersion.parse(installed) ?: return false
        val candidate = SemanticVersion.parse(available) ?: return false
        if (current.isDevelopmentBuild || candidate.isPrerelease) return false
        return candidate > current
    }
}
